characteristic_notify/web_app.js

```diff
@@ -0,0 +1,140 @@
+/**
+ *  Copyright 2020 Google LLC
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+// The test service defined in device_code.js
+const testService = '02fc549a-244c-11eb-adc1-0242ac120002';
+// The test characteristic defined in device_code.js
+const testCharacteristic = '02fc549a-244c-11eb-adc1-0242ac120002';
+const requiredNumUpdates = 100;
+
+/**
+ * Load the device code to the Espruino IDE.
+ */
+function loadEspruinoDeviceCode() {
+  fetch('device_code.js').then(response => response.text()).then(data => {
+    let url = 'http://www.espruino.com/webide?code=' + encodeURIComponent(data);
+    window.open(url, '_window');
+  });
+}
+
+function onGattDisconnected(evt) {
+  const device = evt.target;
+  logInfo(`Disconnected from GATT on device ${device.name}.`);
+}
+
+async function startTest() {
+  clearStatus();
+  logInfo('Starting test');
+
+  $('btn_start_test').disabled = true;
+  $('btn_load_code').disabled = true;
+  let gattServer = undefined;
+  let notifyCharacteristic = undefined;
+  let updateNum = 0;
+  let lastValue = null;
+
+  const resetTest = async () => {
+    if (notifyCharacteristic) {
+      await notifyCharacteristic.stopNotifications();
+    }
+    if (gattServer && gattServer.connected) {
+      logInfo('Disconnecting from GATT.');
+      gattServer.disconnect();
+    }
+    $('btn_start_test').disabled = false;
+    $('btn_load_code').disabled = false;
+  }
+
+  const checkCharacteristicValue = (value) => {
+    if (value < 1 || value > 9999) {
```

Comment:
0 is a valid value when the counter rolls over.

Reply:
Thanks. Fixed to be zero. I also manually tested successful wrapping by lowering the wrap value to 30 to make sure my wrap logic was correct.